### PR TITLE
Add MultivariateNormalDiag distribution

### PR DESCRIPTION
**Pull request by samuela**

### CODE REVIEW

**Comment by alicanb** [comment]
I'm not sure we actually need this. This can be done with Normal only (+ Independent if you're strict about `event_shape`)

**Comment by fritzo** [comment]
@alicanb is correct, this can be accomplished via `Independent(Normal(loc, scale_diag), 1)`. Maybe that should be an example in the `Normal` or `Independent` docs?

**Comment by alicanb** [comment]
@fritzo there is an example in `Independent` docs. (written by you actually 😄 )

**Comment by zou3519** [comment]
Have you tried benchmarking @samuela?

**Comment by samuela** [comment]
@zou3519 I have not. I assumed that this PR was sort of doomed based on the discussion.

**Comment by fritzo** [comment]
I think there will be negligible performance benefit, since `Independent` simply calls its base dist
```py
    def rsample(self, sample_shape=torch.Size()):
        return self.base_dist.rsample(sample_shape)

    def log_prob(self, value):
        log_prob = self.base_dist.log_prob(value)
        return _sum_rightmost(log_prob, self.reinterpreted_batch_ndims)
```
Also IMO it is cleaner to provide a general `Independent` wrapper to create diagonal distributions of any base type, rather than implement `DiagNormal`, `DiagGamma`, `DiagCauchy`, `DiagBeta` etc. If you find `Independent(Normal(...), ...)` cumbersome, we could add a `Normal.independent(...)` method as we [have](http://docs.pyro.ai/en/0.2.1-release/_modules/pyro/distributions/torch_distribution.html#TorchDistributionMixin.independent) in Pyro.

File: torch/distributions/multivariate_normal_diag.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        if scale_diag.shape[-1:] != event_shape:
+            raise ValueError("scale_diag must be a batch of vectors with shape {}".format(event_shape))
+
+        try:
+            self.loc, self.scale_diag = torch.broadcast_tensors(loc, scale_diag)
+        except RuntimeError:
+            raise ValueError("Incompatible batch shapes: loc {}, scale_diag {}"
+                             .format(loc.shape, scale_diag.shape))
+        batch_shape = self.loc.shape[:-1]
+        super(LowRankMultivariateNormal, self).__init__(batch_shape, event_shape,
```
Comment (kashif) [inline torch/distributions/multivariate_normal_diag.py]: typo here?

Comment (samuela) [reply torch/distributions/multivariate_normal_diag.py]: ah, yes. will fix

**Comment by fritzo** [comment]
Can we please replace this complex PR in favor of @samuela's simple wrapper?
```py
def MultivariateNormalDiag(loc, scale_diag):
    if loc.dim() < 1:
        raise ValueError("loc must be at least one-dimensional.")
    return Independent(Normal(loc, scale_diag), 1)
```

**Comment by JakobHavtorn** [comment]
As far as I am aware, another reason for wanting a separate diagonal class is the `torch.distributions.kl_divergence` raises a `NotImplementedError` when called on a diagonal Normal defined with `Independent(Normal(loc, scale_diag), 1)`. It works fine with `MultivariateNormal` but this can be memory intensive.

**Comment by fritzo** [comment]
> torch.distributions.kl_divergence raises a NotImplementedError

Note this works with pyro.distributions. If you'd like this behavior with torch.distributions, feel free to move upstream the Pyro implementations of [kl_divergence(Independent,Independent)](https://github.com/pyro-ppl/pyro/blob/dev/pyro/distributions/kl.py#L19) and [kl_divergence(Independent(Normal),MultivariateNormal)](https://github.com/pyro-ppl/pyro/blob/dev/pyro/distributions/kl.py#L32).

@samuela What do you think of Pyro's abbreviated syntax [.to_event()](https://github.com/pyro-ppl/pyro/blob/c8e9668/pyro/distributions/torch_distribution.py#L108), used as
```py
Normal(...).to_event(1) = Independent(Normal(...), 1)
```
If there is demand we can move that upstream to torch.distributions.Distribution. I think that would be a better solution than a custom new class that works for only one distribution.

**Comment by samuela** [comment]
@fritzo  I've started using jax now actually, so I haven't been following this issue much. `.to_event(...)` looks interesting although I would worry about offering to ways to accomplish the same thing.

**Comment by JakobHavtorn** [comment]
@fritzo Thank you, it is completely straightforward to get that working. I'm new to contributing to PyTorch, are there any specific things I need to comply with to make a PR (and somewhere to read about this)?

**Comment by lqf96** [comment]
Is the current `log_prob` behavior something we want? Because the following implementation:

```python
def log_prob(self, value):
        log_prob = self.base_dist.log_prob(value)
        return _sum_rightmost(log_prob, self.reinterpreted_batch_ndims)
```

is clearly different from the PDF described in the [Tensorflow Distributions document](https://www.tensorflow.org/probability/api_docs/python/tfp/distributions/MultivariateNormalDiag):

```
pdf(x; loc, scale) = exp(-0.5 ||y||**2) / Z,
y = inv(scale) @ (x - loc),
Z = (2 pi)**(0.5 k) |det(scale)|,
```

I think many people expects the log of PDF when they are calling `log_prob`, but clearly PyTorch is not returning the log of PDF for `MultivariateNormalDiag` on `Independent(Normal(loc, scale), 1)`.

**Pull request closed without merging**